### ISSUE TITLE
Exercise suggestions

### ISSUE THREAD

**Issue by chessboards** [head-first-csharp/fourth-edition#35]
I have a proposal that I believe will drastically improve many of the exercises in the book. Of course we all have our different learning styles, however I personally find exercises such as magnets & fill in the blank pools to be a bit of drudgery. There is a book I have read long ago called "Python Crash Course" by Eric Matthes that captured my attention towards how exercises were completely in the hands of the reader. To reinforce learned concepts, the reader was tasked to create a program based on loose guidelines from the very beginning. I think these if they were to replace the two aforementioned exercise types would greatly improve knowledge retention and fun.

Also, at least in my physical copy there are a few pool exercises where some blanks have been printed in such a way that it appears there are two instead of one. Maybe it's just luck of the draw. It appears most notably on page 639.

**Comment by narniaexpert** [comment]
I don't have a problem with puzzle based exercises. The difficulty I have found was transitioning from the print copy to eBook. The print book has sharpen your pencil exercises and fill in the blanks which are good to keep readers engaged. What about if we are reading on a device where we can's write on the screen? What if someone doesn't want to write in their book (like a borrowed resource)? My suggestion is to have interactive webpage based versions of puzzles that can be accessed from the github repository or from an eBook hyperlink.

**Comment by andrewstellman** [comment]
@chessboards We've done a lot of experimenting over the years, and while some readers like you—especially more advanced ones who know a few languages already—sometimes skip the puzzles, a lot of our readers have told us that they're actually really valuable. For a lot of folks, especially new learners, creating a program based on loose guidelines feels like being thrown in the deep end—they basically stare at the page and have no idea what to do. We do recognize that the pool puzzles in particular aren't necessarily for everyone, which is why we told readers that they're optional. But the magnets in particular really help people who are still somewhat struggling with the syntax to put together code that works without having to deal with a "clean slate" scenario, and that can be a really important stepping stone towards being able to write code from the ground up.

Thanks for pointing out the issue on page 639, though—you're right, there's a weird gap in the underscore. It's not a print problem, it's actually my fault. If you're curious, it's an artifact of how I designed the puzzle in Adobe InDesign (which we use to lay out the pages). Basically, I took the exercise solution, copied and pasted it, and turned the text white, so the gap the blank in `if( dingo ________ >2){` is made by a white-colored "g" dipping below the line. I'll figure out how to fix this in future printings.

@narniaexpert Would it help to have PDF downloads of all of the blank exercises that you can print out?

**Comment by Delgathar** [comment]
Yes.  The magnets are very frustrating and back in previous editions, I skipped them.  This time I'm forcing myself to do them and there is definitely a benefit for me. Even though I've yet to get a puzzle correct.

The negative about learning is the pre-generated code given.  I don't learn from "adding this main method to your program" much.  I keep having to go back to read how the ReadKey() works.  But of course, it has to be presented the first time.   Just saying less is more.  The good thing about the magnet and pool puzzles is because the ENTIRE PROGRAM is on me.  Even a little bit of seed code can solve a lot of problems before I create them.  The more of the program that is up to me, the more likely I'll create some simple logic error.

I did have to get an updated tablet with a pen and get program that would allow me  to write in my pdfs.  It helped a lot.  

Since DRM will often make an ebook unprintable or editable, one has to either circumvent this somehow or make due.  Doing the Chapter 8 downloadable exercise now, so these are good.

Squid seems to be a pretty good program for writing in pdfs.  Adobe and Evernote are very clunky so I don't recommend them.

**Comment by andrewstellman** [comment]
@Delgathar Thank you so much for sharing that! That is super, super helpful—it's so valuable to us authors to hear both positive and negative experiences people have with our books. You're absolute right about DRM making things really annoying when it comes to editing, I wish we had more control over that in eBook editions like ones you buy for Kindle.

Let me talk it over with Jenny and see if we can do something to make it easier to do the puzzles—especially the Code Magnets and Pool Puzzle ones—because it's really important to us that our readers like you have the best experience possible with our book.
